programs/drift/src/state/lp_pool.rs

```diff
@@ -0,0 +1,1462 @@
+            .safe_div(QUOTE_PRECISION_I128)?;
+
+        let c_post_in: i128 = gamma_covar_error_post_in
+            .safe_mul(post_notional_errors[0])?
+            .safe_div(2)?
+            .safe_div(QUOTE_PRECISION_I128)?;
+        let c_post_out = gamma_covar_error_post_out
+            .safe_mul(post_notional_errors[1])?
+            .safe_div(2)?
+            .safe_div(QUOTE_PRECISION_I128)?;
+
+        let in_fee = c_post_in
+            .safe_sub(c_pre_in)?
+            .safe_mul(PERCENTAGE_PRECISION_I128)?
+            .safe_div(trade_notional)?
+            .safe_mul(QUOTE_PRECISION_I128)?
+            .safe_div(self.last_aum.cast::<i128>()?)?;
+        let out_fee = c_post_out
+            .safe_sub(c_pre_out)?
+            .safe_mul(PERCENTAGE_PRECISION_I128)?
+            .safe_div(trade_notional)?
+            .safe_mul(QUOTE_PRECISION_I128)?
+            .safe_div(self.last_aum.cast::<i128>()?)?;
+
+        Ok((in_fee, out_fee))
+    }
+
+    pub fn get_linear_fee_execution(
+        &self,
+        trade_notional: i128,
+        kappa_execution: u64,
+        xi: u8,
+        spot_depth: u128,
+    ) -> DriftResult<i128> {
+        let trade_ratio: i128 = trade_notional
+            .abs()
+            .safe_mul(PERCENTAGE_PRECISION_I128)?
+            .safe_div(spot_depth.cast::<i128>()?)?;
+
+        trade_ratio
+            .safe_mul(kappa_execution.safe_mul(xi as u64)?.cast::<i128>()?)?
+            .safe_div(PERCENTAGE_PRECISION_I128)
+    }
+
+    pub fn get_quadratic_fee_execution(
+        &self,
+        trade_notional: i128,
+        kappa_execution: u64,
+        xi: u8,
+        spot_depth: u128,
+    ) -> DriftResult<i128> {
+        let scaled_abs_trade_notional = trade_notional
+            .abs()
+            .safe_mul(PERCENTAGE_PRECISION_I128)?
+            .safe_div(spot_depth.cast::<i128>()?)?;
+
+        kappa_execution
+            .cast::<i128>()?
+            .safe_mul(xi.safe_mul(xi)?.cast::<i128>()?)?
+            .safe_mul(scaled_abs_trade_notional.safe_mul(scaled_abs_trade_notional)?)?
+            .safe_div(PERCENTAGE_PRECISION_I128)?
+            .safe_div(PERCENTAGE_PRECISION_I128)
+    }
+
+    /// returns fee in PERCENTAGE_PRECISION
+    pub fn get_swap_fees(
+        &self,
+        in_spot_market: &SpotMarket,
+        in_oracle_price: i64,
+        in_constituent: &Constituent,
+        in_amount: i128,
+        in_target_weight: i64,
+        out_spot_market: Option<&SpotMarket>,
+        out_oracle_price: Option<i64>,
+        out_constituent: Option<&Constituent>,
+        out_target_weight: Option<i64>,
+        correlation: i64,
+    ) -> DriftResult<(i128, i128)> {
+        msg!("in target weight: {}", in_target_weight);
+        msg!("out target weight: {:?}", out_target_weight);
+
+        let notional_trade_size =
+            in_constituent.get_notional(in_oracle_price, in_spot_market, in_amount, false)?;
+        let out_amount = if out_oracle_price.is_some() {
```

Comment:
if let Some(out_oracle_price) = out_oracle_price {

}

it's cleaner

gotta do this everywhere in this fn